File: 72-edit-distance/edit-distance.go
import "fmt"
func minDistance(word1 string, word2 string) int {
    m, n := len(word1), len(word2)
    if n == m && word1 == word2 {
        return 0
    }
    
    // DPテーブル（スライス）を作成
    dp := make([][]int, m+1)
    for i := range dp {
        dp[i] = make([]int, n+1)
    }
    // ベースケース：空の文字列への変換
    for i := 0; i<=m ; i++ {
        dp[i][0] = i
    }
    for j := 0; j<=n; j++ {
        dp[0][j] = j
    }

    for i:=1; i<=m; i++{
        for j:=1; j<=n; j++ {
            if word1[i-1] == word2[j-1] {
                dp[i][j] = dp[i-1][j-1]
            } else {
                dp[i][j] = min(
                    dp[i-1][j-1]+1, // 置換
                    dp[i][j-1]+1,// 挿入
                    dp[i-1][j]+1,// 削除
                )
            }
            // fmt.Println(i,j, dp)
        }
    }
    return dp[m][n]
}

File: 72-edit-distance/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        if word1 == word2:
            return 0
        dp = [[0]*(n+1) for _ in range(m+1)]
        # base case
        for i in range(m + 1):
            dp[i][0] = i  # 削除のみ
        for j in range(n + 1):
            dp[0][j] = j  # 挿入のみ


        for i in range(1,m+1):
            for j in range(1, n+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    replace = dp[i-1][j-1] + 1
                    insert = dp[i][j-1] + 1
                    delete = dp[i-1][j] + 1
                    dp[i][j] = min(replace, insert, delete)
        return dp[m][n]